hive-metastore/src/main/java/org/apache/iceberg/hive/legacy/LegacyHiveSchemaUtils.java

```diff
@@ -0,0 +1,691 @@
+      if (oldSchema.getType().equals(Schema.Type.ENUM) && newSchema.getType().equals(Schema.Type.STRING)) {
+        return false;
+      }
+
+      LOG.info("Schema is evolved. The old schema type is: {}, the new schema type is: {}",
+          oldSchema.getType(), newSchema.getType());
+      return true;
+    }
+
+    switch (newSchema.getType()) {
+      case BOOLEAN:
+      case BYTES:
+      case DOUBLE:
+      case FLOAT:
+      case INT:
+      case LONG:
+      case STRING:
+      case FIXED:
+        return false;
+      case ENUM:
+        return oldSchema.getEnumSymbols().size() != newSchema.getEnumSymbols().size();
+
+      case RECORD:
+        return isRecordSchemaEvolved(oldSchema, newSchema);
+
+      case MAP:
+        return isSchemaEvolved(oldSchema.getValueType(), newSchema.getValueType());
+
+      case ARRAY:
+        return isSchemaEvolved(oldSchema.getElementType(), newSchema.getElementType());
+
+      case UNION:
+        boolean isBothNullableType =
+            AvroSchemaUtil.isOptionSchema(oldSchema) && AvroSchemaUtil.isOptionSchema(newSchema);
+
+        if (!isBothNullableType) {
+          throw new IllegalArgumentException(
+              "We do not support UNION type " + "except nullable field in schema: " + newSchema.toString(true));
+        }
+
+        Schema oldNonNullType = AvroSchemaUtil.fromOption(oldSchema);
+        Schema newNonNullType = AvroSchemaUtil.fromOption(newSchema);
+
+        return isSchemaEvolved(oldNonNullType, newNonNullType);
+      default:
+        throw new IllegalArgumentException(
+            "Unsupported Avro type " + newSchema.getType() + " in new schema: " + newSchema.toString(true));
+    }
+  }
+
+  private static boolean isPartitioned(@Nonnull Table tableOrView) {
+    Preconditions.checkNotNull(tableOrView, "tableOrView cannot be null");
+
+    List<FieldSchema> partitionColumns = getPartitionCols(tableOrView);
+
+    if (partitionColumns == null) {
+      return false;
+    }
+
+    return partitionColumns.size() != 0;
+  }
+
+  private static List<FieldSchema> getPartitionCols(@Nonnull Table tableOrView) {
+    Preconditions.checkNotNull(tableOrView, "tableOrView cannot be null");
+
+    List<FieldSchema> partKeys = tableOrView.getPartitionKeys();
+    if (partKeys == null) {
+      partKeys = new ArrayList<>();
+      tableOrView.setPartitionKeys(partKeys);
+    }
+
+    return partKeys;
+  }
+
+  private static void mergeAndAppendField(Schema.Field sourceField, Schema.Field metaDataField,
+      SchemaBuilder.FieldAssembler<Schema> fieldAssembler) {
+    Schema.Type actualMetaDataFieldType = extractActualTypeIfFieldIsNullableType(metaDataField.schema()).getType();
+    boolean isSameType = sourceField.schema().getType().equals(actualMetaDataFieldType);
+    boolean isMetaDataFieldEnumType = metaDataField.schema().getType().equals(Schema.Type.ENUM);
+    boolean useMetaDataType = isSameType || isMetaDataFieldEnumType;
```

Comment:
If our contract is to use the Hive schema types, we should not borrow the Enum types from the supplemental schema as well.

Reply:
What is the value of making a special case to use MetaDataType if is the same type? Why can we just use the Hive type by default, and then we would not have to make this condition altogether? (especially after removing `isMetaDataFieldEnumType` as well.